pkg/operator/controllers/muo/muo_controller.go

```diff
@@ -0,0 +1,135 @@
+	pullSpec string = "rh.srep.muo.deploy.pullspec"
+)
+
+type Reconciler struct {
+	log           *logrus.Entry
+	arocli        aroclient.Interface
+	kubernetescli kubernetes.Interface
+	dh            dynamichelper.Interface
+	deployer      *deployer
+
+	readinessPollTime time.Duration
+	readinessTimeout  time.Duration
+}
+
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, kubernetescli kubernetes.Interface, dh dynamichelper.Interface) *Reconciler {
+	return &Reconciler{
+		log: log,
+
+		arocli:        arocli,
+		kubernetescli: kubernetescli,
+		dh:            dh,
+		deployer:      newDeployer(kubernetescli, dh),
+
+		readinessPollTime: 10 * time.Second,
+		readinessTimeout:  5 * time.Minute,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.OperatorFlags.GetSimpleBoolean(enabled) {
+		// controller is disabled
+		return reconcile.Result{}, nil
+	}
+
+	managed := instance.Spec.OperatorFlags.GetWithDefault(managed, "")
+
+	// If enabled and managed=true, install MUO
+	// If enabled and managed=false, remove the MUO deployment
+	// If enabled and managed is missing, do nothing
+	if strings.EqualFold(managed, "true") {
+		imagePullspec, ext := instance.Spec.OperatorFlags[pullSpec]
+		if !ext {
+			return reconcile.Result{}, errors.New("missing pullspec")
+		}
+
+		err = r.deployer.CreateOrUpdate(ctx, imagePullspec)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Check that MUO has become ready, wait up to readinessTimeout (default 5min)
+		timeoutCtx, cancel := context.WithTimeout(ctx, r.readinessTimeout)
+		defer cancel()
+
+		err := wait.PollImmediateUntil(r.readinessPollTime, func() (bool, error) {
+			return r.deployer.IsReady(ctx)
+		}, timeoutCtx.Done())
+		if err != nil {
+			r.log.Error("Managed Upgrade Operator deployment timed out on Ready")
+			return reconcile.Result{}, err
+		}
+	} else if strings.EqualFold(managed, "false") {
+		errs := r.deployer.Remove(ctx)
+		if len(errs) != 0 {
+			r.log.Error("Error removing MUO:")
+			for _, e := range errs {
+				r.log.Error(e)
+			}
+			// just return one of the errors
+			return reconcile.Result{}, errs[0]
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// SetupWithManager setup our manager
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	aroClusterPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetName() == arov1alpha1.SingletonClusterName
+	})
+
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}, builder.WithPredicates(aroClusterPredicate))
+
+	resources, err := r.deployer.resources("")
+	if err != nil {
+		return err
+	}
+
+	for _, i := range resources {
+		o, ok := i.(client.Object)
+		if ok {
+			builder.Owns(o)
+		}
+	}
```

Comment:
1. I'm not sure that this will work. `deployer.resources` returns a slice of `kruntime.Object`, not a specific type. Will controller-runtime be able to figure out a specific type/kind from it?
2. Where do we set ownership for these objects we deploy?

---

We need careful testing here:

1. We need to make sure that objects we deploy have the ownership set properly.
2. We need to make sure that making a change on objects we deploy - will trigger the reconciler.

Reply:
In the logs when I tested it, I got this:

```
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: /, Kind="
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=RoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=Role"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=ClusterRoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=Role"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: /v1, Kind=Namespace"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=RoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=ClusterRole"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=RoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=Role"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: /v1, Kind=ServiceAccount"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=RoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: apps/v1, Kind=Deployment"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=Role"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: /v1, Kind=ConfigMap"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
time="2022-02-17T06:54:44Z" level=info msg="Starting Controller" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:173" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster
time="2022-02-17T06:54:44Z" level=info msg="Starting workers" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:207" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster worker count=1
```

I think that means it's working?

Reply:
Okay, I've now added setting ownership of objects we control, and a test for it.